x-pack/legacy/plugins/index_management/public/components/mappings_editor/reducer.ts

```diff
@@ -61,29 +63,85 @@ export type Action =
 
 export type Dispatch = (action: Action) => void;
 
+export const addFieldToState = (field: Field, state: State): State => {
+  const id = getUniqueId();
+  const { fieldToAddFieldTo } = state.documentFields;
+  const addToRootLevel = fieldToAddFieldTo === undefined;
+  const parentField = addToRootLevel ? undefined : state.fields.byId[fieldToAddFieldTo!];
+  const isMultiField = parentField ? parentField.canHaveMultiFields : false;
+
+  const rootLevelFields = addToRootLevel
+    ? [...state.fields.rootLevelFields, id]
+    : [...state.fields.rootLevelFields];
+  const nestedDepth =
+    parentField && parentField.canHaveChildFields ? parentField.nestedDepth + 1 : 0;
+  const maxNestedDepth = Math.max(state.fields.maxNestedDepth, nestedDepth);
+  const { name } = field;
+  const path = parentField ? `${parentField.path}.${name}` : name;
+
+  state.fields.byId[id] = {
+    id,
+    parentId: fieldToAddFieldTo,
+    isMultiField,
+    source: field,
+    path,
+    nestedDepth,
+    ...getFieldMeta(field),
+  };
+
+  if (parentField) {
+    const childFields = parentField.childFields || [];
+
+    // Update parent field with new children
+    state.fields.byId[fieldToAddFieldTo!] = {
+      ...parentField,
+      childFields: [...childFields, id],
+      hasChildFields: parentField.canHaveChildFields ? true : false,
+      hasMultiFields: parentField.canHaveMultiFields ? true : false,
+      isExpanded: true,
+    };
+
+    // We _also_ need to make a copy of the parent "childFields"
+    // array to force a re-render in the view.
+    if (parentField.parentId) {
```

Comment:
I'm not entirely clear on why this is need - as I understand it the `byId` object is a flat mapping on the tree so not sure what we gain by copying the array in this fashion.

Reply:
You are right, the logic had some complexity because of some memoization down the component tree. I updated the logic and removed the need of those non-intuitive array updates. Thanks for pointing this!